doc/contributing/erm-guidelines.md

```diff
@@ -0,0 +1,329 @@
+sought to introduce ERM capabilities into the `fs.mkdtemp` API such that a
+temporary directory could be created and automatically disposed of when the
+scope in which it was created exited. However, the existing implementation of
+the `fs.mkdtemp` API returns a string value that cannot be made disposable.
+There are also sync, callback, and promise-based variations of the existing
+API that further complicate the situation.
+
+In the initial proposal, the `fs.mkdtemp` API was changed to return an object
+that implements the `Symbol.dispose` method but only if a specific option is
+provided. This would mean that the return value of the API would become
+polymorphic, returning different types based on how it was called. This adds
+a lot of complexity to the API and makes it difficult to reason about the
+return value. It also makes it difficult to programmatically detect whether
+the version of the API being used supports ERM capabilities or not. That is,
+`fs.mkdtemp('...', { disposable: true })` would act differently in older versions
+of Node.js than in newer versions with no way to detect this at runtime other
+than to inspect the return value.
+
+Some APIs that already return object that can be made disposable do not have
+this kind of issue. For example, the `setTimeout` API in Node.js returns an
+object that implements the `Symbol.dispose` method. This change was made without
+much fanfare because the return value of the API was already an object.
+
+So, some APIs can be made disposable easily without any issues while others
+require more thought and consideration. The following guidelines can help
+when introducing ERM capabilities into existing APIs:
+
+1. Avoid polymorphic return values: If an API already returns a value that
+   can be made disposable and it makes sense to make it disposable, do so. Do
+   not, however, make the return value polymorphic determined by an option
+   passed into the API.
+2. Introduce new API variants that are ERM capable: If an existing API
+   cannot be made disposable without changing the return type or making it
+   polymorphic, consider introducing a new API variant that is ERM capable.
+   For example, `fs.mkdtempDisposable` could be introduced to return a
+   disposable object while the existing `fs.mkdtemp` API continues to return
+   a string. Yes, it means more APIs to maintain but it avoids the complexity
+   and confusion of polymorphic return values.
+3. If adding a new API variant is not ideal, remember that changing the
+   return type of an existing API is a breaking change.
+4. When an existing API signature does not lend itself easily to supporting making
+    the return value disposable and a new API needs to be introduced, it is worth
+    considering whether the existing API should be deprecated in favor of the new.
+    Deprecation is never a decision to be taken lightly, however, as it can have major
+    ecosystem impact.
```

Comment:
```suggestion
4. When an existing API signature does not lend itself easily to supporting making
   the return value disposable and a new API needs to be introduced, it is worth
   considering whether the existing API should be deprecated in favor of the new.
   Deprecation is never a decision to be taken lightly, however, as it can have major
   ecosystem impact.
```